### PR TITLE
Develop observation collector

### PR DESCRIPTION
**Pull request by vincentpierre**

### CODE REVIEW

**Comment by chriselion** [comment]
Seems OK. I'm not sure it actually makes the API any smaller though, just moves it to VectorSensor.

**Comment by vincentpierre** [comment]
Yes, but it does simplify the Agent code (since it does not forward the calls to addobservation for every type of observation.

**Comment by chriselion** [comment]
But it also couples the public interface of Agent more closely to VectorSensor, whereas before they were decoupled (and we could theoretically change the implementation of how AddVectorObs worked)

**Comment by chriselion** [comment]
I retract my objections. Make it so!

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 1]
```diff
@@ -539,7 +539,7 @@ void UpdateSensors()
         /// Visual observations are implicitly added from the cameras attached to
         /// the Agent.
         /// </remarks>
-        public virtual void CollectObservations()
+        public virtual void CollectObservations(VectorSensor sensor)
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Agent.cs]: Please update the docstring for this.

File: docs/Learning-Environment-Design-Agents.md [hunk 2]
```diff
@@ -129,7 +129,7 @@ properties to use a continuous vector observation:
 The observation feature vector is a list of floating point numbers, which means
 you must convert any other data types to a float or a list of floats.
 
-The `AddVectorObs` method provides a number of overloads for adding common types
+The `sensor.AddObservation` method provides a number of overloads for adding common types
```
Comment (chriselion) [inline docs/Learning-Environment-Design-Agents.md]:
```suggestion
The `VectorSensor.AddObservation` method provides a number of overloads for adding common types
```

File: docs/Learning-Environment-Design-Agents.md [hunk 3]
```diff
@@ -144,27 +144,27 @@ the feature vector. The following code example illustrates how to add.
...
 enum CarriedItems { Sword, Shield, Bow, LastItem }
 const int NUM_ITEM_TYPES = (int)CarriedItems.LastItem;
 
-public override void CollectObservations()
+public override void CollectObservations(VectorSensor sensor)
 {
     // The first argument is the selection index; the second is the
     // number of possibilities
-    AddVectorObs((int)currentItem, NUM_ITEM_TYPES);
+    sensor.AddObservation((int)currentItem, NUM_ITEM_TYPES);
```
Comment (chriselion) [inline docs/Learning-Environment-Design-Agents.md]:
```suggestion
    sensor.AddOneHotObservation((int)currentItem, NUM_ITEM_TYPES);
```

File: docs/Learning-Environment-Design-Agents.md [hunk 4]
```diff
@@ -144,27 +144,27 @@ the feature vector. The following code example illustrates how to add.
...
     for (int ci = 0; ci < (int)CarriedItems.LastItem; ci++)
     {
-        AddVectorObs((int)currentItem == ci ? 1.0f : 0.0f);
+        sensor.AddObservation((int)currentItem == ci ? 1.0f : 0.0f);
     }
 }
 ```
 
-`AddVectorObs` also provides a two-argument version as a shortcut for _one-hot_
+`sensor.AddObservation` also provides a two-argument version as a shortcut for _one-hot_
```
Comment (chriselion) [inline docs/Learning-Environment-Design-Agents.md]:
```suggestion
`VectorSensor.AddObservation` also provides a two-argument version as a shortcut for _one-hot_
```

File: docs/Getting-Started-with-Balance-Ball.md [hunk 5]
```diff
@@ -105,7 +105,7 @@ This means that the feature
 vector containing the Agent's observations contains eight elements: the `x` and
 `z` components of the agent cube's rotation and the `x`, `y`, and `z` components
 of the ball's relative position and velocity. (The observation values are
-defined in the Agent's `CollectObservations()` function.)
+defined in the Agent's `CollectObservations(VectorSensor sensor)` function.)
```
Comment (chriselion) [inline docs/Getting-Started-with-Balance-Ball.md]:
```suggestion
defined in the Agent's `CollectObservations(VectorSensor sensor)` method.)
```

File: docs/Migrating.md [hunk 6]
```diff
@@ -7,8 +7,15 @@ The versions can be found in
 
 # Migrating
 
+## Migrating from 0.14 to latest
 
-## Migrating from 0.13 to latest
+### Important changes
+* The `CollectObservations()` virtual method now needs as input a `VectorSensor` sensor as argument.
```
Comment (chriselion) [inline docs/Migrating.md]:
```suggestion
* The `Agent.CollectObservations()` virtual method now takes as input a `VectorSensor` sensor as argument. The `Agent.AddVectorObs()` methods were removed.
```

File: docs/Migrating.md [hunk 7]
```diff
@@ -7,8 +7,15 @@ The versions can be found in
...
+
+### Steps to Migrate
+* Replace your implementation of `CollectObservations()` with `CollectObservations(VectorSensor sensor)`. In addition, replace all calls to `AddVectorObs()` with `sensor.AddObservation()` or `sensor.AddOneHotObservation()` on the `VectorSensor` passed as argument.
```
Comment (chriselion) [inline docs/Migrating.md]:
```suggestion
* Replace your Agent's implementation of `CollectObservations()` with `CollectObservations(VectorSensor sensor)`. In addition, replace all calls to `AddVectorObs()` with `sensor.AddObservation()` or `sensor.AddOneHotObservation()` on the `VectorSensor` passed as argument.
```

File: com.unity.ml-agents/Runtime/Agent.cs (hunk 1 above)
Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Agent.cs]: Let me know what you think

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Agent.cs]: Sounds good. I'm not sure about the syntax on cref, but we can sort that out when we start building the package docs.

File: com.unity.ml-agents/Runtime/Sensor/VectorSensor.cs [hunk 8]
```diff
@@ -1,7 +1,8 @@
 using System.Collections.Generic;
 using UnityEngine;
+using MLAgents.Sensor;
 
-namespace MLAgents.Sensor
+namespace MLAgents
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Sensor/VectorSensor.cs]: What's the reason for this change? Should we (roughly) keep the namespaces 1:1 with the subdirectories?

Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensor/VectorSensor.cs]: Yes but now the user would have to import both MLAgents and MLAgents.Sensors when implementing an Agent that has a CollectObservation implementation

Comment (vincentpierre) [reply com.unity.ml-agents/Runtime/Sensor/VectorSensor.cs]: All right. Removed the `MLAgents.Sensor` namespace alltogether

**Comment by chriselion** [comment]
:shipit:
